### ISSUE TITLE
itertools.cycle fails for generators

### ISSUE THREAD

**Issue by stinos** [micropython/micropython-lib#41]
The cycle implementation does not make a copy of the elements iterated over and as such only works for proper sequences like lists, but not for all types of iterables:

```
from itertools import cycle

def reverse(l):
    while l:
        yield l.pop()

for x in cycle(reverse([1, 2, 3])):
    print(x)
```

output:

```
3
2
1

```

Proposed fix: just copy the equivalent python implementation?

```
def cycle(iterable):
    saved = []
    for element in iterable:
        yield element
        saved.append(element)
    while saved:
        for element in saved:
              yield element
```

**Comment by pfalcon** [comment]
> Proposed fix: just copy the equivalent python implementation?

Do you mean copy from CPython stdlib? Simple copy is never simple copy, there're also licensing issues to account for.

Also, the above implementation clearly bloats memory in case simple sequence is passed, so I'd suggest to add check for that and avoid copying in that case (I know there're bunch of corner cases, but let's resolve them on being hit). Feel free to submit a patch.

**Comment by stinos** [comment]
> Do you mean copy from CPython stdlib?

No I just meant copying python implementation as shown from https://docs.python.org/3/library/itertools.html#itertools.cycle, not the actual C implementation.

**Comment by dpgeorge** [comment]
That definition almost mandates the given implementation.  And, as it says, it's expected to use lots of memory.

**Comment by stinos** [comment]
Fixed with #42
